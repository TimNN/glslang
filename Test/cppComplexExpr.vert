#define ON1
#define ON2

float sum = 0.0;

void main()
{
#if defined(ON1) && (defined(OFF) || defined(ON2))
//yes
    sum += 1.0;
#endif

#if !defined(ON1) || (defined(OFF) || (!defined(OFF2) && defined(ON2)))
//yes
    sum += 20.0;
#endif

#if defined(ON1) && (defined(OFF) || !defined(ON2))
//no
    sum += 0.1;
#endif

#if !defined(ON1) || (defined(OFF) || !defined(OFF2) && !defined(ON2))
//no
    sum += 0.2;
#endif

#if !defined(ON1) || !defined(OFF) || defined(ON2) && defined(OFF2)
//yes
    sum += 300.0;
#endif

#if (!defined(ON1) || !defined(OFF) || defined(ON2)) && defined(OFF2)
//no
    sum += 0.4;
#endif

// sum should be 321.0
    gl_Position = vec4(sum);
}

#define ADD(a, b) a + b + ((a) + ((b)));

float foo()
{
    return ADD(gl_Position.xyxwx, 3.0)  // ERROR, should be this line number
    return ADD(gl_Position.y, 3.0)
}

#if 0
// ERROR, EOF