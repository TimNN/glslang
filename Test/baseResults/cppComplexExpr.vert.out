ERROR: 0:46: 'xyxwx' : illegal vector field selection 
ERROR: 0:46: 'xyxwx' : illegal vector field selection 
ERROR: 0:66: '#define' : Macro redefined; different substitutions: BIG
ERROR: 0:79: '' : missing #endif 
ERROR: 4 compilation errors.  No code generated.

ERROR: node is still EOpNull!
0:4  Sequence
0:4    move second child to first child (highp float)
0:4      'sum' (highp float)
0:4      Constant:
0:4        0.000000
0:6  Function Definition: main( (void)
0:6    Function Parameters: 
0:10    Sequence
0:10      add second child into first child (highp float)
0:10        'sum' (highp float)
0:10        Constant:
0:10          1.000000
0:15      add second child into first child (highp float)
0:15        'sum' (highp float)
0:15        Constant:
0:15          20.000000
0:30      add second child into first child (highp float)
0:30        'sum' (highp float)
0:30        Constant:
0:30          300.000000
0:39      move second child to first child (highp 4-component vector of float)
0:39        'gl_Position' (gl_Position highp 4-component vector of float)
0:39        Construct vec4 (highp 4-component vector of float)
0:39          'sum' (highp float)
0:44  Function Definition: foo( (highp float)
0:44    Function Parameters: 
0:46    Sequence
0:46      Branch: Return with expression
0:46        add (highp float)
0:46          add (highp float)
0:46            direct index (highp float)
0:46              'gl_Position' (gl_Position highp 4-component vector of float)
0:46              Constant:
0:46                0 (const int)
0:46            Constant:
0:46              3.000000
0:46          add (highp float)
0:46            direct index (highp float)
0:46              'gl_Position' (gl_Position highp 4-component vector of float)
0:46              Constant:
0:46                0 (const int)
0:46            Constant:
0:46              3.000000
0:47      Branch: Return with expression
0:47        add (highp float)
0:47          add (highp float)
0:47            direct index (highp float)
0:47              'gl_Position' (gl_Position highp 4-component vector of float)
0:47              Constant:
0:47                1 (const int)
0:47            Constant:
0:47              3.000000
0:47          add (highp float)
0:47            direct index (highp float)
0:47              'gl_Position' (gl_Position highp 4-component vector of float)
0:47              Constant:
0:47                1 (const int)
0:47            Constant:
0:47              3.000000
0:?   Linker Objects
0:?     'sum' (highp float)
0:?     'gl_VertexID' (gl_VertexId highp int)
0:?     'gl_InstanceID' (gl_InstanceId highp int)

